src/canard.c

```diff
@@ -1,7 +1,919 @@
+
+        if (ins->should_accept(ins, &data_type_signature, data_type_id, transfer_type, source_node_id))
+        {
+            rxstate = canardRxStateTraversal(ins, transfer_descriptor);
+            rxstate->calculated_crc = crcAddSignature(0xFFFFU, data_type_signature);
+        }
+    }
+    else
+    {
+        rxstate = canardFindRxState(ins->rx_states, transfer_descriptor);
+    }
+
+    if (rxstate == NULL)
+    {
+        return;
+    }
+
+    // Resolving the state flags:
+    const bool not_initialized = (rxstate->timestamp_usec == 0) ? true : false;
+    const bool tid_timed_out = (timestamp_usec - rxstate->timestamp_usec) > TRANSFER_TIMEOUT_USEC;
+    const bool first_frame = IS_START_OF_TRANSFER(tail_byte);
+    const bool not_previous_tid =
+        computeForwardDistance(rxstate->transfer_id, TRANSFER_ID_FROM_TAIL_BYTE(tail_byte)) > 1;
+
+    bool need_restart =
+        (not_initialized) ||
+        (tid_timed_out) ||
+        (first_frame && not_previous_tid);
+
+    if (need_restart)
+    {
+        rxstate->transfer_id = TRANSFER_ID_FROM_TAIL_BYTE(tail_byte);
+        rxstate->next_toggle = 0;
+        canardReleaseStatePayload(ins, rxstate);
+        if (!IS_START_OF_TRANSFER(tail_byte)) // missed the first frame
+        {
+            rxstate->transfer_id += 1;
+            if (rxstate->transfer_id >= 32)
+            {
+                rxstate->transfer_id = 0;
+            }
+            return;
+        }
+    }
+
+    if (IS_START_OF_TRANSFER(tail_byte) && IS_END_OF_TRANSFER(tail_byte)) // single frame transfer
+    {
+        rxstate->timestamp_usec = timestamp_usec;
+        CanardRxTransfer rxtransfer = {
+            .payload_head = frame->data,
+            .payload_len = frame->data_len - 1,
+            .data_type_id = data_type_id,
+            .transfer_type = transfer_type,
+            .priority = priority,
+            .source_node_id = source_node_id
+        };
+        ins->on_reception(ins, &rxtransfer);
+
+        canardPrepareForNextTransfer(rxstate);
+        return;
+    }
+
+    if (TOGGLE_BIT(tail_byte) != rxstate->next_toggle)
+    {
+        return; // wrong toggle
+    }
+
+    if (TRANSFER_ID_FROM_TAIL_BYTE(tail_byte) != rxstate->transfer_id)
+    {
+        return; // unexpected tid
+    }
+
+    if (IS_START_OF_TRANSFER(tail_byte) && !IS_END_OF_TRANSFER(tail_byte)) // beginning of multi frame transfer
+    { // take off the crc and store the payload
+        rxstate->timestamp_usec = timestamp_usec;
+        canardBufferBlockPushBytes(&ins->allocator, rxstate, frame->data + 2, frame->data_len - 3);
+        rxstate->payload_crc = ((uint16_t)frame->data[0] << 8) | ((uint16_t)frame->data[1]);
+        rxstate->calculated_crc = crcAdd(rxstate->calculated_crc, frame->data + 2, frame->data_len - 3);
+    }
+    else if (!IS_START_OF_TRANSFER(tail_byte) && !IS_END_OF_TRANSFER(tail_byte))
+    {
+        canardBufferBlockPushBytes(&ins->allocator, rxstate, frame->data, frame->data_len - 1);
```

Comment:
Missing error handling here and in the similar invocation above. Also see my comments on `canardBufferBlockPushBytes()`.